File: java/Dson-Core/src/main/java/cn/wjybxx/dson/types/ExtDateTime.java
package cn.wjybxx.dson.types;

import cn.wjybxx.dson.DsonLites;
import cn.wjybxx.dson.internal.DsonInternals;

import javax.annotation.concurrent.Immutable;
import java.time.*;
import java.time.format.DateTimeFormatter;

/**
 * 日期时间
 *
 * @author wjybxx
 * date - 2023/6/17
 */
@Immutable
public final class ExtDateTime {

    public static final byte MASK_NONE = 0;
    public static final byte MASK_DATE = 1;
    public static final byte MASK_TIME = 1 << 1;
    public static final byte MASK_OFFSET = 1 << 2;

    public static final byte MASK_DATETIME = MASK_DATE | MASK_TIME;
    public static final byte MASK_DATETIME_OFFSET = MASK_DATE | MASK_TIME | MASK_OFFSET;
    public static final byte MASK_ALL = MASK_DATETIME_OFFSET;

    /** 纪元时间-秒 */
    private final long seconds;
    /** 纪元时间的纳秒部分 */
    private final int nanos;
    /** 时区偏移-秒 */
    private final int offset;
    /** 哪些字段有效 */
    private final byte enables;

    /**
     * 该接口慎用，通常我们需要精确到毫秒
     *
     * @param seconds 纪元时间-秒
     */
    public ExtDateTime(long seconds) {
        this(seconds, 0, 0, MASK_DATETIME);
    }

    public ExtDateTime(long seconds, int nanos, int offset, byte enables) {
        if ((enables & MASK_ALL) != enables) {
            throw new IllegalArgumentException("invalid enables: " + enables);
        }
        if (seconds != 0 && !DsonInternals.isAnySet(enables, MASK_DATETIME)) {
            throw new IllegalArgumentException("date and time are disabled, but seconds is not 0");
        }
        if (nanos != 0 && !DsonInternals.isSet(enables, MASK_TIME)) {
            throw new IllegalArgumentException("time is disabled, but nanos is not 0");
        }
        if (offset != 0 && !DsonInternals.isSet(enables, MASK_OFFSET)) {
            throw new IllegalArgumentException("offset is disabled, but the value is not 0");
        }
        Timestamp.validateNanos(nanos);
        this.seconds = seconds;
        this.nanos = nanos;
        this.offset = offset;
        this.enables = enables;
    }

    public static ExtDateTime ofDateTime(LocalDateTime localDateTime) {
        long epochSecond = localDateTime.toEpochSecond(ZoneOffset.UTC);
        int nanos = localDateTime.getNano();
        return new ExtDateTime(epochSecond, nanos, 0, MASK_DATETIME);
    }

    public LocalDateTime toDateTime() {
        return LocalDateTime.ofEpochSecond(seconds, nanos, ZoneOffset.UTC);
    }

    public static ExtDateTime ofInstant(Instant instant) {
        return new ExtDateTime(instant.getEpochSecond(), instant.getNano(), 0, MASK_DATETIME);
    }

    public ExtDateTime withOffset(int offset) {
        return new ExtDateTime(seconds, nanos, offset, (byte) (enables | MASK_OFFSET));
    }

    public ExtDateTime withoutOffset() {
        return new ExtDateTime(seconds, nanos, 0, (byte) (enables & MASK_DATETIME));
    }

    // region

    public long getSeconds() {
        return seconds;
    }

    public int getNanos() {
        return nanos;
    }

    public int getOffset() {
        return offset;
    }

    public byte getEnables() {
        return enables;
    }

    public boolean hasDate() {
        return DsonInternals.isSet(enables, MASK_DATE);
    }

    public boolean hasTime() {
        return DsonInternals.isSet(enables, MASK_TIME);
    }

    public boolean hasOffset() {
        return DsonInternals.isSet(enables, MASK_OFFSET);
    }

    public boolean hasFields(byte mask) {
        return DsonInternals.isSet(enables, mask);
    }

    public boolean canBeAbbreviated() {
        return nanos == 0 && (enables == MASK_DATETIME);
    }

    public boolean canConvertNanosToMillis() {
        return (nanos % 1000_000) == 0;
    }

    public int convertNanosToMillis() {
        return nanos / 1000_000;
    }
    // endregion

    //region equals

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        ExtDateTime offsetDateTime = (ExtDateTime) o;

        if (seconds != offsetDateTime.seconds) return false;
        if (nanos != offsetDateTime.nanos) return false;
        if (offset != offsetDateTime.offset) return false;
        return enables == offsetDateTime.enables;
    }

    @Override
    public int hashCode() {
        int result = (int) (seconds ^ (seconds >>> 32));
        result = 31 * result + nanos;
        result = 31 * result + offset;
        result = 31 * result + enables;
        return result;
    }

    // endregion

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("ExtDateTime{");
        if (hasDate()) {
            sb.append("date: '").append(formatDate(seconds));
        }
        if (hasTime()) {
            if (hasDate()) {
                sb.append(", ");
            }
            sb.append("time: '").append(formatTime(seconds));
        }
        if (nanos != 0) {
            sb.append(", ");
            if (canConvertNanosToMillis()) {
                sb.append("millis: ").append(convertNanosToMillis());
            } else {
                sb.append("nanos: ").append(nanos);
            }
        }
        if (hasOffset()) {
            sb.append(", ");
            sb.append("offset: '").append(formatOffset(offset))
                    .append("'");
        }
        return sb.append('}')
                .toString();
    }

    // region parse/format

    /** @return 固定格式 yyyy-MM-dd */
    public static String formatDate(long epochSecond) {
        return LocalDateTime.ofEpochSecond(epochSecond, 0, ZoneOffset.UTC)
                .toLocalDate()
                .toString();
    }

    /** @return 固定格式 HH:mm:ss */
    public static String formatTime(long epochSecond) {
        return LocalDateTime.ofEpochSecond(epochSecond, 1, ZoneOffset.UTC)
                .toLocalTime()
                .toString()
                .substring(0, 8);
    }

    /** @return 固定格式 yyyy-MM-dd'T'HH:mm:ss */
    public static String formatDateTime(long epochSecond) {
        return formatDate(epochSecond) + "T" + formatTime(epochSecond);
    }

    /** @param dateString 限定格式 yyyy-MM-dd */
    public static LocalDate parseDate(String dateString) {
//        if (dateString.length() != 10) throw new IllegalArgumentException("invalid dateString " + dateString);
        return LocalDate.parse(dateString, DateTimeFormatter.ISO_DATE);
    }

    /** @param timeString 限定格式 HH:mm:ss */
    public static LocalTime parseTime(String timeString) {
        if (timeString.length() != 8) throw new IllegalArgumentException("invalid timeString " + timeString);
        return LocalTime.parse(timeString, DateTimeFormatter.ISO_TIME);
    }

    /** @param timeString 限定格式 yyyy-MM-dd'T'HH:mm:ss */
    public static LocalDateTime parseDateTime(String timeString) {
        return LocalDateTime.parse(timeString, DateTimeFormatter.ISO_DATE_TIME);
    }

    /**
     * Z, ±HH:mm, ±HH:mm:ss
     */
    public static String formatOffset(int offsetSeconds) {
        if (offsetSeconds == 0) {
            return "Z";
        }
        String pre = offsetSeconds < 0 ? "-" : "+";
        return pre + LocalTime.ofSecondOfDay(Math.abs(offsetSeconds)).toString();
    }

    /**
     * Z, ±H, ±HH, ±HH:mm, ±HH:mm:ss
     */
    public static int parseOffset(String offsetString) {
        return switch (offsetString.length()) {
            case 1, 2, 3, 6, 9 -> ZoneOffset.of(offsetString).getTotalSeconds();
            default -> throw new IllegalArgumentException("invalid offset string " + offsetString);
        };
    }

    // endregion

    // region 常量
    public static final String NAMES_DATE = "date";
    public static final String NAMES_TIME = "time";
    public static final String NAMES_MILLIS = "millis";

    public static final String NAMES_SECONDS = "seconds";
    public static final String NAMES_NANOS = "nanos";
    public static final String NAMES_OFFSET = "offset";
    public static final String NAMES_ENABLES = "enables";

    public static final int NUMBERS_SECONDS = DsonLites.makeFullNumberZeroIdep(0);
    public static final int NUMBERS_NANOS = DsonLites.makeFullNumberZeroIdep(1);
    public static final int NUMBERS_OFFSET = DsonLites.makeFullNumberZeroIdep(2);
    public static final int NUMBERS_ENABLES = DsonLites.makeFullNumberZeroIdep(3);

    // endregion
}

File: csharp/Wjybxx.Dson.Core/src/Types/ExtDateTime.cs
#region LICENSE

//  Copyright 2023-2024 wjybxx(845740757@qq.com)
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
// 
//      http://www.apache.org/licenses/LICENSE-2.0
// 
//  Unless required by applicable law or agreed to iBn writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#endregion

using System;
using System.Globalization;
using Wjybxx.Commons;
using Wjybxx.Dson.Internal;

namespace Wjybxx.Dson.Types
{
/// <summary>
/// 日期时间
/// 为提高辨识度，我们命名为'ExtDateTime'
/// </summary>
public readonly struct ExtDateTime : IEquatable<ExtDateTime>
{
    public const byte MaskNone = 0;
    public const byte MaskDate = 1;
    public const byte MaskTime = 1 << 1;
    public const byte MaskOffset = 1 << 2;

    public const byte MaskDatetime = MaskDate | MaskTime;
    public const byte MaskDatetimeOffset = MaskDate | MaskTime | MaskOffset;
    public const byte MaskAll = MaskDatetimeOffset;

    /** 纪元时间-秒 */
    public long Seconds { get; }
    /** 纪元时间的纳秒部分 */
    public int Nanos { get; }
    /** 时区偏移-秒 */
    public int Offset { get; }
    /** 哪些字段有效 */
    public byte Enables { get; }

    /// <summary>
    /// 该接口慎用，通常我们需要精确到毫秒
    /// </summary>
    /// <param name="seconds">纪元时间秒时间戳</param>
    public ExtDateTime(long seconds)
        : this(seconds, 0, 0, MaskDatetime) {
    }

    /// <summary>
    /// 
    /// </summary>
    /// <param name="seconds">纪元时间秒时间戳</param>
    /// <param name="nanos">时间戳的纳秒部分</param>
    /// <param name="offset">时区偏移量--秒</param>
    /// <param name="enables">启用的字段信息；只有有效的字段才会被保存(序列化)</param>
    public ExtDateTime(long seconds, int nanos, int offset, byte enables) {
        if ((enables & MaskAll) != enables) {
            throw new ArgumentException("invalid enables: " + enables);
        }
        if (seconds != 0 && !DsonInternals.IsAnySet(enables, MaskDatetime)) {
            throw new ArgumentException("date and time are disabled, but seconds is not 0");
        }
        if (nanos != 0 && !DsonInternals.IsSet(enables, MaskTime)) {
            throw new ArgumentException("time is disabled, but nanos is not 0");
        }
        if (offset != 0 && !DsonInternals.IsSet(enables, MaskOffset)) {
            throw new ArgumentException("offset is disabled, but the value is not 0");
        }
        Timestamp.ValidateNanos(nanos);
        this.Seconds = seconds;
        this.Nanos = nanos;
        this.Offset = offset;
        this.Enables = enables;
    }

    private const int NanosPerTick = 100;
    private const int TicksPerSecond = 10_000_000;

    public static ExtDateTime OfDateTime(in DateTime dateTime) {
        long totalTicks = dateTime.Subtract(DateTime.UnixEpoch).Ticks;
        long seconds = totalTicks / TicksPerSecond; // totalSeconds
        int nanos = (int)(totalTicks % TicksPerSecond * NanosPerTick); // remainNanos -- 取余避免越界
        return new ExtDateTime(seconds, nanos, 0, MaskDatetime);
    }

    public DateTime ToDateTime() {
        long totalTicks = Seconds * TicksPerSecond + Nanos / NanosPerTick;
        return DateTime.UnixEpoch.Add(new TimeSpan(totalTicks));
    }

    public ExtDateTime WithOffset(int offset) {
        return new ExtDateTime(Seconds, Nanos, offset, (byte)(Enables | MaskOffset));
    }

    public ExtDateTime WithoutOffset() {
        return new ExtDateTime(Seconds, Nanos, 0, (byte)(Enables & MaskDatetime));
    }

    #region props

    public bool HasDate => DsonInternals.IsSet(Enables, MaskDate);

    public bool HasTime => DsonInternals.IsSet(Enables, MaskTime);

    public bool HasOffset => DsonInternals.IsSet(Enables, MaskOffset);

    public bool HasFields(byte mask) {
        return DsonInternals.IsSet(Enables, mask);
    }

    /** 是否可以缩写 */
    public bool CanBeAbbreviated() {
        return Nanos == 0 && (Enables == MaskDatetime);
    }

    /** 纳秒部分是否可转为毫秒 -- 纳秒恰为整毫秒时返回true */
    public bool CanConvertNanosToMillis() {
        return (Nanos % 1000_000) == 0;
    }

    /** 将纳秒部分换行为毫秒 */
    public int ConvertNanosToMillis() {
        return Nanos / 1000_000;
    }

    # endregion

    #region equals

    public bool Equals(ExtDateTime other) {
        return Seconds == other.Seconds && Nanos == other.Nanos && Offset == other.Offset && Enables == other.Enables;
    }

    public override bool Equals(object? obj) {
        return obj is ExtDateTime other && Equals(other);
    }

    public override int GetHashCode() {
        return HashCode.Combine(Seconds, Nanos, Offset, Enables);
    }

    public static bool operator ==(ExtDateTime left, ExtDateTime right) {
        return left.Equals(right);
    }

    public static bool operator !=(ExtDateTime left, ExtDateTime right) {
        return !left.Equals(right);
    }

    #endregion

    public override string ToString() {
        return $"{nameof(Seconds)}: {Seconds}, {nameof(Nanos)}: {Nanos}, {nameof(Offset)}: {Offset}, {nameof(Enables)}: {Enables}";
    }

    #region 解析

    public static DateTime ParseDateTime(string datetimeString) {
        return DateTime.ParseExact(datetimeString, "yyyy-MM-ddTHH:mm:ss", CultureInfo.InvariantCulture);
    }

    /** 为避免dotnet5的兼容性问题，我们返回DateTime */
    public static DateTime ParseDate(string dateString) {
        return DateTime.ParseExact(dateString + "T00:00:00", "yyyy-MM-ddTHH:mm:ss", CultureInfo.InvariantCulture);
    }

    /** 为避免dotnet的兼容性问题，我们返回总秒数 */
    public static int ParseTime(string timeString) {
        return DatetimeUtil.ParseTime2(timeString);
    }

    /// <summary>
    /// 格式化日期时间为ISO-8601格式
    /// 固定为:<code>yyyy-MM-ddTHH:mm:ss</code>
    /// </summary>
    public static string FormatDateTime(long seconds) {
        return DateTime.UnixEpoch.AddSeconds(seconds).ToString("s");
    }

    /// <summary>
    /// 格式化日期为ISO-8601格式
    /// 固定为:<code>"yyyy-MM-dd"</code>格式
    /// </summary>
    public static string FormatDate(long epochSeconds) {
        DateTime dateTime = DateTime.UnixEpoch.AddSeconds(epochSeconds);
        string fullString = dateTime.ToString("s");
        return fullString.Substring(0, fullString.IndexOf('T'));
    }

    /// <summary>
    /// 格式化时间为ISO-8601格式
    /// 固定为:<code>HH:mm:ss</code>格式
    /// </summary>
    public static string FormatTime(long epochSeconds) {
        DateTime dateTime = DateTime.UnixEpoch.AddSeconds(epochSeconds);
        int secondOfDay = DatetimeUtil.ToSecondOfDay(dateTime.Hour, dateTime.Minute, dateTime.Second);
        return DatetimeUtil.FormatTime2(secondOfDay);
    }

    /// <summary>
    /// 解析时区偏移
    /// Z, ±H, ±HH, ±H:mm, ±HH:mm, ±HH:mm:ss
    /// </summary>
    /// <param name="offsetString"></param>
    /// <returns></returns>
    public static int ParseOffset(string offsetString) {
        return DatetimeUtil.ParseOffset(offsetString);
    }

    public static string FormatOffset(int offsetSeconds) {
        return DatetimeUtil.FormatOffset(offsetSeconds);
    }

    #endregion

    #region 常量

    public const string NamesDate = "date";
    public const string NamesTime = "time";
    public const string NamesMillis = "millis";

    public const string NamesSeconds = "seconds";
    public const string NamesNanos = "nanos";
    public const string NamesOffset = "offset";
    public const string NamesEnables = "enables";

    public static readonly FieldNumber NumbersSeconds = FieldNumber.OfLnumber(0);
    public static readonly FieldNumber NumbersNanos = FieldNumber.OfLnumber(1);
    public static readonly FieldNumber NumbersOffset = FieldNumber.OfLnumber(2);
    public static readonly FieldNumber NumbersEnables = FieldNumber.OfLnumber(3);

    #endregion
}
}